### PR TITLE
Add support for validating GitHub Release Asset checksums

### PR DESCRIPTION
**Pull request by josh-padnick**
Fetch is used to download both GitHub files and Release Assets. When downloading a Release Asset, it's helpful to verify the SHA256 or SHA512 hash of the release asset (the "checksum") so that you can confirm the release asset hasn't been changed from the time you first downloaded it. This PR adds optional support for doing that.

This use case came up for me when I was downloading a third-party health checker tool using Fetch. Without verifying the checksum, the downloaded tool could be changed without anyone knowing, a potential security issue given that we do not control the third-party repo. Using bash to validate the checksum is an option, but fetch support leads to slightly cleaner code and avoids platform compatibility issues. 

In addition, I opted to make a backwards-incompatible change to simplify the UX. Previously, fetch allowed users to download more than one Release Asset in a single call. Once I implemented this PR, the fetch interface quickly became unwieldy with so many flags to pass in when using multiple release assets. 

It occurred to me that (a) downloading multiple release assets is itself probably a less likely use case, and (b) users can simply call fetch multiple times, once for each release asset they want to download. Therefore, I opted for the UX simplicity with no loss of functionality but at the cost of backwards-incompatible change.

Other updates:
- Slight cleanup to the README

### CODE REVIEW

**Comment by brikis98** [comment]
A bummer, I just filed this as a GitHub issue with the "newbie" label. It would make for a good trial project :)

Will review shortly.

**Comment by josh-padnick** [comment]
Ah, it did occur to me that it felt a little trial-projectish, but I wanted to actually use it in my Confluent setup. Thanks for the forthcoming review!

File: checksum_test.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	"io/ioutil"
+)
+
+const SAMPLE_RELEASE_ASSET_GITHUB_REPO_URL ="https://github.com/gruntwork-io/health-checker"
+const SAMPLE_RELEASE_ASSET_VERSION="v0.0.2"
+const SAMPLE_RELEASE_ASSET_NAME="health-checker_linux_amd64"
+
+// Checksums can be computed by running "shasum -a [256|512] /path/to/file" on any UNIX system
+const SAMPLE_RELEASE_ASSET_CHECKSUM_SHA256="4314590d802760c29a532e2ef22689d4656d184b3daa63f96bc8b8f76f5d22f0"
+const SAMPLE_RELEASE_ASSET_CHECKSUM_SHA512="28d9e487c1001e3c28d915c9edd3ed37632f10b923bd94d4d9ac6d28c0af659abbe2456da167763d51def2182fef01c3f73c67edf527d4ed1389a28ba10db332"
```
Comment (brikis98) [inline checksum_test.go]: Where do you users get these checksums from? If it's from the `/releases` page, then anyone who can replace the asset with a fake one could also replace the checksum with a fake one. To be truly useful, wouldn't we have to publish these checksums in some totally separate location? Or are you making the assumption that when someone goes to get the checksum value, they are verifying that the code is valid, and this merely ensures no one can pull the carpet out from under them later?

File: main.go [hunk 2]
```diff
@@ -235,32 +261,33 @@ func downloadSourcePaths(sourcePaths []string, destPath string, githubRepo GitHu
...
-// Download the specified binary files that were uploaded as release assets to the specified GitHub release
-func downloadReleaseAssets(releaseAssets []string, destPath string, githubRepo GitHubRepo, latestTag string) error {
-	if len(releaseAssets) == 0 {
-		return nil
+// Download the specified binary file that was uploaded as a release asset to the specified GitHub release.
+// Returns the path where the release asset was downloaded.
+func downloadReleaseAsset(assetName string, destPath string, githubRepo GitHubRepo, tag string) (string, error) {
+	var assetPath string
+
+	if assetName == "" {
```
Comment (brikis98) [inline main.go]: Should this be an error?

File: README.md [hunk 3]
```diff
@@ -64,8 +65,11 @@ The supported options are:
   the `/folder` path and all files below it). By default, all files are downloaded from the repo unless `--source-path`
   or `--release-asset` is specified. This option can be specified more than once.
 - `--release-asset` (**Optional**): The name of a release asset--that is, a binary uploaded to a [GitHub
-  Release](https://help.github.com/articles/creating-releases/)--to download. This option can be specified more than
-  once. It only works with the `--tag` option.
+  Release](https://help.github.com/articles/creating-releases/)--to download. It only works with the `--tag` option.
+- `--release-asset-checksum` (**Optional**): The checksum that a release asset should have. Fetch will fail if this value
+  is non-empty and does not match the checksum computed by Fetch.
+- `--release-asset-checksum-algo` (**Optional**): The algorithm fetch will use to compute a checksum of the release asset.
```
[the remark this answers is not in the document]
Comment (josh-padnick) [reply README.md]: Do you mean the `--modules` pattern of `gruntwork-install`? Well, can't you just use Fetch to download the exact commit you want to achieve what you're suggesting?

File: checksum.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	"os"
+	"io"
+	"hash"
+	"encoding/hex"
+)
+
+func verifyChecksumOfReleaseAsset(assetPath, checksum, algorithm string) *FetchError {
+	computedChecksum, err := computeChecksum(assetPath, algorithm)
+	if err != nil {
+		return newError(ERROR_WHILE_COMPUTING_CHECKSUM, err.Error())
```
[the remark this answers is not in the document]
Comment (josh-padnick) [reply checksum.go]: I hear you. This was an early Golang tool I wrote and I didn't want to refactor all the errors throughout the code so I stuck with the existing idiom.

File: checksum.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	if computedChecksum != checksum {
+		return newError(CHECKSUM_DOES_NOT_MATCH, fmt.Sprintf("Expected to receive checksum value %s, but instead got %s for Release Asset at %s", computedChecksum, checksum, assetPath))
+	}
+
+	fmt.Printf("Checksum matches!")
```
[the remark this answers is not in the document]
Comment (josh-padnick) [reply checksum.go]: I was surprised about that as well. Again, I didn't want to invest time in updating those kinds of things so I stuck with the existing idiom.

File: checksum.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	"encoding/hex"
+)
+
+func verifyChecksumOfReleaseAsset(assetPath, checksum, algorithm string) *FetchError {
+	computedChecksum, err := computeChecksum(assetPath, algorithm)
+	if err != nil {
+		return newError(ERROR_WHILE_COMPUTING_CHECKSUM, err.Error())
+	}
+	if computedChecksum != checksum {
+		return newError(CHECKSUM_DOES_NOT_MATCH, fmt.Sprintf("Expected to receive checksum value %s, but instead got %s for Release Asset at %s", computedChecksum, checksum, assetPath))
```
[the remark this answers is not in the document]
Comment (josh-padnick) [reply checksum.go]: That's a good idea. I'll update now.

[the remark this answers is not in the document]
Comment (josh-padnick) [reply checksum.go]: 3bb9834

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (brikis98) [reply README.md]: Yes, `--modules`. We typically use tags for specifying the version we want. Should our recommendation be that users specify both the tag and the commit ID?

File: checksum.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	fmt.Printf("Checksum matches!")
+
+	return nil
+}
+
+func computeChecksum(filePath string, algorithm string) (string, error) {
+	var checksum string
```
[the remark this answers is not in the document]
Comment (josh-padnick) [reply checksum.go]: I sometimes like to declare the return var at the top, but your suggestion is probably a little cleaner. updated.

File: main.go (hunk 2 above)
Comment (josh-padnick) [reply main.go]: No. This code got updated when I changed `assetNames` from a `[]string` to a `string`. That being said, I think the error handling could generally be improved here, as discussed above.

File: checksum_test.go (hunk 1 above)
Comment (josh-padnick) [reply checksum_test.go]: Exactly. When I went to download a third-party health checker for Kafka, I realized how risky (and yet common!) including such a library is. If an attacker wanted to exploit a library that many people use, one way is to find a way into someone's GitHub account (leaked GitHub token?), leave the commits unaltered and update only the release asset (where in most cases no checksum at all is published).

In this case, I'm making the assumption that at the time you download the code, you have an opportunity to look around the repo and conclude for yourself that it appears not to be compromised. There's only so much you can do here, of course, but at least if in the future when you've long forgotten about this library, something malicious happens, you'll be notified.

In the end, I suppose it's just more defense in depth.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (josh-padnick) [reply README.md]: Well, keep in mind that fetch allows you to use [tag constrain expressions](https://github.com/gruntwork-io/fetch#tag-constraint-expressions), but I see where you're going with this. Yes, I think we should permit using both a `--tag` and a `--commit-id` param at the same time. Fetch can then error out if the two don't match. I think I'll file a GitHub issue for that for now, though.

[the remark this answers is not in the document]
Comment (josh-padnick) [reply README.md]: https://github.com/gruntwork-io/fetch/issues/35

**Comment by josh-padnick** [comment]
All feedback responded to and tests passing, so merging now.

**Pull request merged**